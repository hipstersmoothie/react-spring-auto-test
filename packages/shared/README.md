# @autorelease/shared

This package contains modules which are **not imported** by the main module.

- `@autorelease/shared/colorMatchers`
  Regular expressions for color strings

- `@autorelease/shared/colors`
  The dictionary of named colors and their numeric values

- `@autorelease/shared/normalizeColor`
  Converts named colors, hexadecimal colors, `rgba` strings, and `hsla` strings
  into their numeric values

- `@autorelease/shared/stringInterpolation`
  Exports the `createStringInterpolator` function (with color support)

## Main module

Import `@autorelease/shared` to access these named imports:

- `Globals` object (updated with its `assign` method)
- `createInterpolator` function
- `is` functions (for runtime type checks)
- Utility hooks (like `useForceUpdate`)
- Every type in `src/types`
