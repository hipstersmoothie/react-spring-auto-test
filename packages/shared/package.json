{
  "name": "@autorelease/shared",
  "version": "9.0.0-canary.809.5.f01ecc2",
  "description": "Globals and shared modules",
  "main": "src/index.ts",
  "scripts": {
    "build": "tsc -p ."
  },
  "publishConfig": {
    "directory": "dist"
  }
}
