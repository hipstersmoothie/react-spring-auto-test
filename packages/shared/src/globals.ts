import {
  InterpolatorConfig,
  FrameRequestCallback,
  InterpolatorArgs,
  FluidValue,
  OneOrMore,
} from './types'
import { ElementType } from 'react'

declare const window: {
  requestAnimationFrame: (cb: (time: number) => void) => number
  cancelAnimationFrame: (id: number) => void
}

//
// Required
//

export let defaultElement: string | ElementType

/** Provide custom logic for native updates */
export let applyAnimatedValues: (node: any, props: any) => boolean | void

/** Provide custom logic for string interpolation */
export let createStringInterpolator: (
  config: InterpolatorConfig<string>
) => (input: number) => string

/** Provide a custom `FrameLoop` instance */
export let frameLoop: {
  active: boolean
  update: (time?: number) => boolean
  onFrame(cb: FrameRequestCallback, next?: boolean): void
  onWrite(cb: FrameRequestCallback): void
  start(spring: any): void
  stop(spring: any): typeof frameLoop
}

//
// Optional
//

export let to: <In, Out>(
  source: OneOrMore<FluidValue>,
  args: InterpolatorArgs<In, Out>
) => FluidValue<Out>

/** Used for timestamps in milliseconds. Read more [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/now) */
export let now = () => Date.now()

declare const performance: { now: () => number }

/** Used to measure frame length. Read more [here](https://developer.mozilla.org/en-US/docs/Web/API/Performance/now) */
export let performanceNow = () => performance.now()

/** Provide custom color names for interpolation */
export let colorNames: { [key: string]: number } | null = null as any

/** Make all animations instant and skip the frameloop entirely */
export let skipAnimation = false

/** Intercept props before they're passed to an animated component */
export let getComponentProps = (props: any) => props

/** Wrap the `style` prop with an animated node */
export let createAnimatedStyle: ((style: any) => any) | null = null as any

/** Wrap the `transform` prop with an animated node */
export let createAnimatedTransform:
  | ((transform: any) => any)
  | null = null as any

export let requestAnimationFrame: typeof window.requestAnimationFrame =
  typeof window !== 'undefined' ? window.requestAnimationFrame : () => -1

export let cancelAnimationFrame: typeof window.cancelAnimationFrame =
  typeof window !== 'undefined' ? window.cancelAnimationFrame : () => {}

//
// Configuration
//

export interface AnimatedGlobals {
  to?: typeof to
  now?: typeof now
  frameLoop?: typeof frameLoop
  colorNames?: typeof colorNames
  skipAnimation?: typeof skipAnimation
  defaultElement?: typeof defaultElement
  performanceNow?: typeof performanceNow
  getComponentProps?: typeof getComponentProps
  applyAnimatedValues?: typeof applyAnimatedValues
  createStringInterpolator?: typeof createStringInterpolator
  createAnimatedTransform?: typeof createAnimatedTransform
  createAnimatedStyle?: typeof createAnimatedStyle
  requestAnimationFrame?: typeof requestAnimationFrame
  cancelAnimationFrame?: typeof cancelAnimationFrame
}

export const assign = (globals: AnimatedGlobals): AnimatedGlobals =>
  ({
    to,
    now,
    frameLoop,
    colorNames,
    skipAnimation,
    defaultElement,
    performanceNow,
    getComponentProps,
    applyAnimatedValues,
    createStringInterpolator,
    createAnimatedTransform,
    createAnimatedStyle,
    requestAnimationFrame,
    cancelAnimationFrame,
  } = Object.assign(
    {
      to,
      now,
      frameLoop,
      colorNames,
      skipAnimation,
      defaultElement,
      performanceNow,
      getComponentProps,
      applyAnimatedValues,
      createStringInterpolator,
      createAnimatedTransform,
      createAnimatedStyle,
      requestAnimationFrame,
      cancelAnimationFrame,
    },
    pluckDefined(globals)
  ))

// Ignore undefined values
function pluckDefined(globals: any) {
  const defined: any = {}
  for (const key in globals) {
    if (globals[key] !== undefined) defined[key] = globals[key]
  }
  return defined
}
