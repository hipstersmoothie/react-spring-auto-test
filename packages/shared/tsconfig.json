{
  "include": ["src"],
  "exclude": ["**/*.test.ts"],
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "declaration": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "inlineSources": true,
    "jsx": "react",
    "lib": ["es2017"],
    "moduleResolution": "node",
    "noEmitOnError": true,
    "noFallthroughCasesInSwitch": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "outDir": "dist",
    "preserveSymlinks": true,
    "sourceMap": true,
    "strict": true,
    "target": "es5",
    "typeRoots": [
      "../../node_modules/@types",
      "./node_modules/@types",
      "./@types"
    ]
  }
}
