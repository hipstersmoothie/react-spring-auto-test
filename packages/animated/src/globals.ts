import { Globals } from 'shared'
import { AnimatedStyle } from './AnimatedStyle'
import { Into } from './Into'

// Sane defaults
Globals.assign({
  to: (source, args) => new Into(source, args),
  createAnimatedStyle: style => new AnimatedStyle(style),
})
