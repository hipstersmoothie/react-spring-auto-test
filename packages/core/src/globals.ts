import { createStringInterpolator } from 'shared/stringInterpolation'
import { FrameLoop } from './FrameLoop'
import { Globals } from 'shared'

Globals.assign({
  frameLoop: new FrameLoop(),
  createStringInterpolator,
  applyAnimatedValues: () => false,
})

export { Globals }

/** Advance all animations forward one frame */
export const update = () => Globals.frameLoop.update()
