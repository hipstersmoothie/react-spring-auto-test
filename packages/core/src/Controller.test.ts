import createMockRaf, { MockRaf } from 'mock-raf'
import * as Globals from 'shared/globals'
import { Controller } from './Controller'
import { FrameLoop } from './FrameLoop'

let mockRaf: MockRaf
beforeEach(() => {
  mockRaf = createMockRaf()
  Globals.assign({
    now: mockRaf.now,
    performanceNow: mockRaf.now,
    requestAnimationFrame: mockRaf.raf,
    cancelAnimationFrame: mockRaf.cancel,
    frameLoop: new FrameLoop(),
  })
})

it('can animate a number', () => {
  const ctrl = new Controller({ x: 0 })
  ctrl.update({ x: 100 }).start()

  const frames = getFrames(ctrl)
  expect(frames).toMatchSnapshot()

  // The first frame should *not* be the from value.
  expect(frames[0]).not.toEqual({ x: 0 })

  // The last frame should be the goal value.
  expect(frames.slice(-1)[0]).toEqual({ x: 100 })
})

it('can animate an array of numbers', () => {
  const config = { precision: 0.005 }
  const ctrl = new Controller<{ x: [number, number] }>({ x: [1, 2], config })
  ctrl.update({ x: [5, 10] }).start()

  const frames = getFrames(ctrl)
  expect(frames).toMatchSnapshot()

  // The last frame should be the goal value.
  expect(frames.slice(-1)[0]).toEqual({ x: [5, 10] })

  // The 2nd value is always ~2x the 1st value (within the defined precision).
  const factors = frames.map(frame => frame.x[1] / frame.x[0])
  expect(
    factors.every(factor => Math.abs(2 - factor) < config.precision)
  ).toBeTruthy()
})

describe('async "to" prop', () => {
  it('acts strangely without the "from" prop', async () => {
    const ctrl = new Controller<{ x: number }>()
    ctrl.start({
      to: async update => {
        // The spring does not exist yet!
        expect(ctrl.get('x')).toBeUndefined()

        // Any values passed here are treated as "from" values,
        // because no "from" prop was ever given.
        update({ x: 1 })
        // Now the spring exists!
        expect(ctrl.get('x')).toBeDefined()
        // But the spring is idle!
        expect(ctrl.get('x').idle).toBeTruthy()

        // This call *will* start an animation!
        update({ x: 2 })
        expect(ctrl.get('x').idle).toBeFalsy()
      },
    })

    // Since we call `update` twice, frames are generated!
    expect(await getAsyncFrames(ctrl)).toMatchSnapshot()
  })
})

/** Collect all frames synchronously */
function getFrames<T extends object>(ctrl: Controller<T>): T[] {
  const frames: any[] = []
  const onFrame = (frame: any) => frames.push(frame)

  let steps = 0
  while (!ctrl.idle) {
    ctrl['_props'].onFrame = onFrame
    mockRaf.step()
    if (++steps > 1e5) {
      throw Error('Infinite loop detected')
    }
  }

  return frames
}

/** Wait one microtask tick between frames */
async function getAsyncFrames<T extends object>(
  ctrl: Controller<T>
): Promise<T[]> {
  const frames: any[] = []
  const onFrame = (frame: any) => frames.push(frame)

  let steps = 0
  while (!ctrl.idle) {
    ctrl['_props'].onFrame = onFrame
    mockRaf.step()
    await Promise.resolve()
    if (++steps > 1e5) {
      throw Error('Infinite loop detected')
    }
  }

  return frames
}
