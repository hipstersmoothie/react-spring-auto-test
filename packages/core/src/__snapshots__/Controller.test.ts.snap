// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async "to" prop acts strangely without the "from" prop 1`] = `
Array [
  Object {
    "x": 1.0326794443028517,
  },
  Object {
    "x": 1.1012936337027077,
  },
  Object {
    "x": 1.1876657320498814,
  },
  Object {
    "x": 1.2801401452733328,
  },
  Object {
    "x": 1.3715121742243184,
  },
  Object {
    "x": 1.4575767171197955,
  },
  Object {
    "x": 1.5361170578451617,
  },
  Object {
    "x": 1.6062054680717361,
  },
  Object {
    "x": 1.6677238686547753,
  },
  Object {
    "x": 1.7210390692738162,
  },
  Object {
    "x": 1.7667859812257725,
  },
  Object {
    "x": 1.8057257371556243,
  },
  Object {
    "x": 1.8386553439187705,
  },
  Object {
    "x": 1.8663524183244127,
  },
  Object {
    "x": 1.8895434886465414,
  },
  Object {
    "x": 1.908887849332717,
  },
  Object {
    "x": 1.924971437373813,
  },
  Object {
    "x": 1.9383069480579156,
  },
  Object {
    "x": 1.9493376351533416,
  },
  Object {
    "x": 1.9584430966505435,
  },
  Object {
    "x": 1.9659459400886563,
  },
  Object {
    "x": 1.9721186285150805,
  },
  Object {
    "x": 1.9771900844749588,
  },
  Object {
    "x": 1.9813518144075553,
  },
  Object {
    "x": 1.984763437341402,
  },
  Object {
    "x": 1.9875575795666838,
  },
  Object {
    "x": 1.9898441449232573,
  },
  Object {
    "x": 1.9917139982178524,
  },
  Object {
    "x": 1.9932421138596101,
  },
  Object {
    "x": 1.994490247769977,
  },
  Object {
    "x": 1.9955091911941107,
  },
  Object {
    "x": 1.996340662391671,
  },
  Object {
    "x": 1.9970188877658095,
  },
  Object {
    "x": 1.9975719187485288,
  },
  Object {
    "x": 1.9980227253010303,
  },
  Object {
    "x": 1.998390101576788,
  },
  Object {
    "x": 1.9986894143427598,
  },
  Object {
    "x": 1.9989332202661931,
  },
  Object {
    "x": 2,
  },
]
`;

exports[`can animate a number 1`] = `
Array [
  Object {
    "x": 3.26794443028518,
  },
  Object {
    "x": 10.129363370270768,
  },
  Object {
    "x": 18.76657320498811,
  },
  Object {
    "x": 28.01401452733326,
  },
  Object {
    "x": 37.15121742243184,
  },
  Object {
    "x": 45.757671711979555,
  },
  Object {
    "x": 53.61170578451619,
  },
  Object {
    "x": 60.62054680717365,
  },
  Object {
    "x": 66.77238686547756,
  },
  Object {
    "x": 72.10390692738164,
  },
  Object {
    "x": 76.67859812257728,
  },
  Object {
    "x": 80.57257371556248,
  },
  Object {
    "x": 83.8655343918771,
  },
  Object {
    "x": 86.63524183244132,
  },
  Object {
    "x": 88.95434886465418,
  },
  Object {
    "x": 90.88878493327174,
  },
  Object {
    "x": 92.49714373738131,
  },
  Object {
    "x": 93.83069480579158,
  },
  Object {
    "x": 94.93376351533416,
  },
  Object {
    "x": 95.84430966505437,
  },
  Object {
    "x": 96.59459400886564,
  },
  Object {
    "x": 97.21186285150807,
  },
  Object {
    "x": 97.71900844749592,
  },
  Object {
    "x": 98.13518144075557,
  },
  Object {
    "x": 98.47634373414024,
  },
  Object {
    "x": 98.75575795666842,
  },
  Object {
    "x": 98.98441449232577,
  },
  Object {
    "x": 99.17139982178526,
  },
  Object {
    "x": 99.32421138596104,
  },
  Object {
    "x": 99.4490247769977,
  },
  Object {
    "x": 99.55091911941105,
  },
  Object {
    "x": 99.63406623916707,
  },
  Object {
    "x": 99.7018887765809,
  },
  Object {
    "x": 99.75719187485285,
  },
  Object {
    "x": 99.80227253010301,
  },
  Object {
    "x": 99.83901015767877,
  },
  Object {
    "x": 99.86894143427595,
  },
  Object {
    "x": 99.8933220266193,
  },
  Object {
    "x": 100,
  },
]
`;

exports[`can animate an array of numbers 1`] = `
Array [
  Object {
    "x": Array [
      1.1307177772114074,
      2.261435554422815,
    ],
  },
  Object {
    "x": Array [
      1.405174534810831,
      2.810349069621662,
    ],
  },
  Object {
    "x": Array [
      1.7506629281995245,
      3.501325856399049,
    ],
  },
  Object {
    "x": Array [
      2.12056058109333,
      4.24112116218666,
    ],
  },
  Object {
    "x": Array [
      2.486048696897273,
      4.972097393794546,
    ],
  },
  Object {
    "x": Array [
      2.8303068684791817,
      5.660613736958363,
    ],
  },
  Object {
    "x": Array [
      3.1444682313806473,
      6.288936462761295,
    ],
  },
  Object {
    "x": Array [
      3.424821872286946,
      6.849643744573892,
    ],
  },
  Object {
    "x": Array [
      3.6708954746191016,
      7.341790949238203,
    ],
  },
  Object {
    "x": Array [
      3.884156277095265,
      7.76831255419053,
    ],
  },
  Object {
    "x": Array [
      4.06714392490309,
      8.13428784980618,
    ],
  },
  Object {
    "x": Array [
      4.222902948622497,
      8.445805897244995,
    ],
  },
  Object {
    "x": Array [
      4.354621375675082,
      8.709242751350164,
    ],
  },
  Object {
    "x": Array [
      4.465409673297651,
      8.930819346595301,
    ],
  },
  Object {
    "x": Array [
      4.558173954586166,
      9.116347909172331,
    ],
  },
  Object {
    "x": Array [
      4.635551397330868,
      9.271102794661736,
    ],
  },
  Object {
    "x": Array [
      4.699885749495252,
      9.399771498990503,
    ],
  },
  Object {
    "x": Array [
      4.753227792231662,
      9.506455584463325,
    ],
  },
  Object {
    "x": Array [
      4.797350540613366,
      9.594701081226733,
    ],
  },
  Object {
    "x": Array [
      4.833772386602174,
      9.667544773204348,
    ],
  },
  Object {
    "x": Array [
      4.863783760354625,
      9.72756752070925,
    ],
  },
  Object {
    "x": Array [
      4.888474514060322,
      9.776949028120644,
    ],
  },
  Object {
    "x": Array [
      4.908760337899835,
      9.81752067579967,
    ],
  },
  Object {
    "x": Array [
      4.925407257630221,
      9.850814515260442,
    ],
  },
  Object {
    "x": Array [
      4.939053749365608,
      9.878107498731216,
    ],
  },
  Object {
    "x": Array [
      4.950230318266735,
      9.90046063653347,
    ],
  },
  Object {
    "x": Array [
      4.959376579693029,
      9.918753159386059,
    ],
  },
  Object {
    "x": Array [
      4.96685599287141,
      9.93371198574282,
    ],
  },
  Object {
    "x": Array [
      4.972968455438441,
      9.945936910876881,
    ],
  },
  Object {
    "x": Array [
      4.977960991079908,
      9.955921982159817,
    ],
  },
  Object {
    "x": Array [
      4.982036764776443,
      9.964073529552886,
    ],
  },
  Object {
    "x": Array [
      4.985362649566684,
      9.970725299133369,
    ],
  },
  Object {
    "x": Array [
      4.988075551063238,
      9.976151102126476,
    ],
  },
  Object {
    "x": Array [
      4.990287674994115,
      9.98057534998823,
    ],
  },
  Object {
    "x": Array [
      4.992090901204121,
      9.984181802408242,
    ],
  },
  Object {
    "x": Array [
      4.993560406307152,
      9.987120812614304,
    ],
  },
  Object {
    "x": Array [
      4.994757657371039,
      9.989515314742079,
    ],
  },
  Object {
    "x": Array [
      4.995171906309277,
      9.991465762129545,
    ],
  },
  Object {
    "x": Array [
      4.995171906309277,
      9.993054193518825,
    ],
  },
  Object {
    "x": Array [
      4.995171906309277,
      9.994347576276779,
    ],
  },
  Object {
    "x": Array [
      5,
      10,
    ],
  },
]
`;
